### PR TITLE
Ethereum connector implementation

### PR DESCRIPTION
**Pull request by rranjan3**
- Ethereum connector implementation
- Truncate worker_id to 32 bytes at enclave_manager
- Fix parse issues of Solidity contracts

Signed-off-by: Rajeev Ranjan <rajeev2.ranjan@intel.com>

### CODE REVIEW

File: blockchain_connector/connector_service.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    uri = options.uri
+    if uri:
+        config["tcf"]["json_rpc_uri"] = uri
+
+    proxy_blockchain = options.blockchain
+    if proxy_blockchain:
+        logging.info("Blockchain to be used as proxy : {}"
+                     .format(proxy_blockchain))
+        if proxy_blockchain == ETHEREUM:
+            logging.info("Going to start Ethereum connector service")
```
Comment (danintel) [inline blockchain_connector/connector_service.py]: Change "Going to" to "About to" ("About to" is more immediate, "Going to" is any time in the future).

File: blockchain_connector/connector_service.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            eth_connector_svc = EthereumConnector(config)
+            eth_connector_svc.start()
+        else:
+            logging.info("Going to start Fabric connector service")
```
Comment (danintel) [inline blockchain_connector/connector_service.py]: Change "Going to" to "About to"

File: blockchain_connector/ethereum/ethereum_connector.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+    def handleEvent(self, event, account, contract):
+        print(f'Event ---> {event}')
+
+    def start(self):
+        logging.info("Ethereum Connector service started")
+
+        # Fetch first worker details from shared KV (via direct API)
+        # and add the worker to block chain.
+        # TODO: Fetch all workers frm shared KV and block chain
```
Comment (danintel) [inline blockchain_connector/ethereum/ethereum_connector.py]: Change "frm" to "from"

File: blockchain_connector/connector_service.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+
+def _parse_config_file(config_file):
+    # Parse config file and return a config dictionary.
+    if config_file:
+        conf_files = [config_file]
+    else:
+        conf_files = [TCFHOME +
+                      "/sdk/avalon_sdk/tcf_connector.toml"]
+    confpaths = ["."]
```
Comment (manojgop) [inline blockchain_connector/connector_service.py]: confpaths  -> conf_paths

File: blockchain_connector/ethereum/ethereum_connector.py [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+
+        work_order_contract_file = tcf_home + "/" + \
+            config["ethereum"]["work_order_contract_file"]
+        work_order_contract_address = \
+            config["ethereum"]["work_order_contract_address"]
+        self._work_order_contract_instance = self._eth_client\
+            .get_contract_instance(
+                work_order_contract_file, work_order_contract_address)
+
+    def _retrieve_registry(self, config):
```
Comment (manojgop) [inline blockchain_connector/ethereum/ethereum_connector.py]: Please remove this function. Retrieve registry is not used in block chain connector

File: blockchain_connector/ethereum/ethereum_connector.py [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+            return None
+        # Retrieve details of the first registry
+        registry_info = worker_registry_list.registry_retrieve(registry_ids[0])
+        logging.info("Registry retrieve response: {}".format(registry_info))
+
+        # Appending id as well to the same tuple
+        return registry_info + (registry_ids[0],)
+
+    def _add_registry_to_chain(self, registry):
+        """
```
Comment (manojgop) [inline blockchain_connector/ethereum/ethereum_connector.py]: Please remove this function. It's not used in block chain connector

File: blockchain_connector/ethereum/ethereum_connector.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+            org_id, uri, sc_addr, app_type_ids).buildTransaction(
+                self._eth_client.get_transaction_params())
+        try:
+            txn_receipt = self._eth_client.execute_transaction(txn_dict)
+        except Execption as e:
+            logging.error("Error adding registry to proxy blockchain"+str(e))
+
+    def _retrieve_first_worker_details(self):
+        """
+        This function retrieves the first workeri from shared kv using
```
Comment (manojgop) [inline blockchain_connector/ethereum/ethereum_connector.py]: workeri -> workeri

File: blockchain_connector/ethereum/ethereum_connector.py [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+    def newListener(self, contract, event, fromBlock='latest'):
+        return contract.events[event].createFilter(fromBlock=fromBlock)
+
+
+class EventProcessor:
+    async def listener(self, eventListener):
+        while True:
+            for event in eventListener.get_new_entries():
+                await self.queue.put(event)
+                print("Evt pushed into Q")
```
Comment (manojgop) [inline blockchain_connector/ethereum/ethereum_connector.py]: remove print. Please use debug log.

File: blockchain_connector/ethereum/ethereum_connector.py [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+            await asyncio.sleep(LISTENER_SLEEP_DURATION)
+
+    async def handler(self, callback, *kargs, **kwargs):
+        while True:
+            event = await self.queue.get()
+            print("Popped event from Q")
```
Comment (manojgop) [inline blockchain_connector/ethereum/ethereum_connector.py]: remove print. Please use debug log.

File: blockchain_connector/ethereum/ethereum_connector.py [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+        await asyncio.gather(*self.listeners)  # infinite loop
+        await self.queue.join()  # this code should never run
+        await self.stop()  # this code should never run
+
+    async def stop(self):
+        for process in self.listeners:
+            process.cancel()
+        for process in self.handlers:
+            process.cancel()
+        print("---exit---")
```
Comment (manojgop) [inline blockchain_connector/ethereum/ethereum_connector.py]: remove print. Please use debug log.

File: sdk/avalon_sdk/ethereum/ethereum_wrapper.py [hunk 11]
```diff
@@ -58,6 +59,9 @@ def __init__(self, config):
             self.__gas_limit = config["ethereum"]["gas_limit"]
             # Amount of Ether you’re willing to pay for every unit of gas
             self.__gas_price = config["ethereum"]["gas_price"]
+            set_solc_version('v0.5.15')
```
Comment (manojgop) [inline sdk/avalon_sdk/ethereum/ethereum_wrapper.py]: Move the version to config file

File: sdk/avalon_sdk/ethereum/ethereum_wrapper.py [hunk 12]
```diff
@@ -193,8 +197,17 @@ def get_contract_instance(self, contract_file_name, contract_address):
         return self.__w3.eth.contract(address=contract_address,
                                       abi=contract_interface["abi"])
 
+    def get_contract_instance_from_json(self, json_file_name,
+                                        contract_address):
```
Comment (manojgop) [inline sdk/avalon_sdk/ethereum/ethereum_wrapper.py]: Add a comment that this json file is compiled solidity file and is not used currently

File: sdk/avalon_sdk/interfaces/work_order_proxy.py [hunk 13]
```diff
@@ -13,10 +13,10 @@
 # limitations under the License.
 
 from abc import ABC, abstractmethod
-from sdk.avalon_sdk.interfaces.work_order import WorkOrder
+from avalon_sdk.interfaces.work_order import WorkOrder
 
 
-class WorkOrderProxy(ABC, WorkOrder):
+class WorkOrderProxy(WorkOrder):
```
Comment (manojgop) [inline sdk/avalon_sdk/interfaces/work_order_proxy.py]: Remove "from abc import ABC, abstractmethod"

File: sdk/avalon_sdk/tcf_connector.toml [hunk 14]
```diff
@@ -34,26 +34,26 @@ worker_registry_contract_file = "sdk/avalon_sdk/ethereum/contracts/WorkerRegistr
...
 #Ethereum account details
-eth_account = "0xDcB780ea30C1041abfcCCE4233c275B9F313eAf4"
+eth_account = "0xefb21e451a2f8c9e25e02ffee969b169a067d343"
 #Ethereum account private key 
 acc_pvt_key = "4F611197A6E82715F4D2446FE015D1667E9C40A351411F3A7300F71F285D01B4"
  
 #Provider for test network. Using Ropsten by default here and Infura as the IAAS.
 #This could be replaced with a url for a Hyperledger Besu client or a Ganache client.
-provider = "https://ropsten.infura.io/v3/9af17e6320414af8a3168cd9fe9d653d"
+provider = "http://10.66.245.80:8545"
```
Comment (manojgop) [inline sdk/avalon_sdk/tcf_connector.toml]: Above comment about Ropsten as default is outdated since Besu is default now ?

Comment (rranjan3) [reply sdk/avalon_sdk/tcf_connector.toml]: Addressed all observation.

File: blockchain_connector/ethereum/ethereum_connector.py (hunk 7 above)
Comment (danintel) [reply blockchain_connector/ethereum/ethereum_connector.py]: Change workeri to worker

**Pull request merged**